lib/eslint.js

```diff
@@ -868,10 +870,11 @@ module.exports = (function() {
                         ruleCreator(ruleContext);
 
                     // add all the node types as listeners
-                    Object.keys(rule).forEach(nodeType => {
-                        api.on(nodeType, timing.enabled
-                            ? timing.time(key, rule[nodeType])
-                            : rule[nodeType]
+                    Object.keys(rule).forEach(selector => {
+                        selectors.add(selector);
+                        api.on(selector, timing.enabled
```

Comment:
Adding every selector could negatively affect performance. This is part of the reason we didn't implement it this way. The current approach has a maximum size of the unique AST node set whereas this has an infinite size. If you look back at past attempts, this is something I brought up frequently as something that needs to be addressed.